website/blog/2025-11-25-node-auto-provisioning/index.md

```diff
@@ -0,0 +1,120 @@
+- Operational Drift: Over time, node pools drifted from their intended purpose. Teams reused them for convenience, leading to noisy neighbor issues and unpredictable performance.
+- Maintenance Complexity: Each node pool had to be patched, upgraded, and monitored separately. This fragmented the upgrade process and increased the risk of downtime.
+
+These challenges weren’t theoretical. They were voiced repeatedly by users across industries. From retail to healthcare to fintech, the message was clear: managing complex infrastructure should not be a prerequisite for using Kubernetes.
+
+## Engineering the Invisible: How We Built Node Auto Provisioning
+
+The AKS engineering team set out to flip the model. What if users didn’t have to define infrastructure at all? What if the platform could provision compute dynamically, based on workload needs?
+
+Enter Node Auto Provisioning, built on top of [Karpenter](https://karpenter.sh/), a CNCF project designed for just-in-time compute provisioning.
+
+### Key Design Principles
+
+**SKU-less by Default**: Users no longer need to specify exact Virtual Machine(VM) sizes. Instead, they define workload requirements (CPU, memory, GPU, etc.), VM families and the platform selects the best-fit VMs. 
+**Mixed VM Size Autoscaling**: Traditional cluster autoscalers scale pre-existing node pools with only one VM size per pool. With NAP, users can automatically scale up single-instance VMs of mixed sizes (SKUs) 
+**Capacity-Aware Scheduling**: NAP is able to react to capacity constraints of certain VM size or regions, by selecting other available VM sizes. 
+**Workload Consolidation**: Inspired by Karpenter’s deprovisioning controller, NAP supports automatic consolidation. If workloads are spread thin across many nodes, the system can reschedule them onto fewer  nodes or even replace nodes with cheaper options, reducing cost and improving efficiency.
+**Customer Preferences, Not Prescriptions**: Customers can still express preferences like Spot vs. On-Demand, preferred VM families - but they don’t have to manage the details.
+**Enforced Resource Requests**: To make intelligent provisioning decisions, we require workloads to define resource requests and limits. This is enforced via admission webhooks, ensuring consistency and predictability.
+**Seamless Upgrades**: We designed NAP to work with AKS’s existing upgrade mechanisms, including node image updates and maintenance windows. Customers can opt into auto-consolidation during maintenance windows to minimize disruption.
+**Simplified Compute Management**: the self-hosted Karpenter project requires additional efforts around bootstrap token rotation and helm charts, which Node Auto Provisioning takes on the responsibility for. As a managed add-on, NAP brings even more simplicity to the Karpenter experience so you can focus on your workloads and not infrastructure. 
```

Comment:
The formatting here is messed up (not rendering as a list) + it is fairly wordy.